deals/deals.go

```diff
@@ -225,6 +234,195 @@ func (m *Module) Watch(ctx context.Context, proposals []cid.Cid) (<-chan DealInf
+	if err != nil {
+		log.Errorf("getting deal info: %v", err)
+		deletePending()
+		return
+	}
+	if info.State != storagemarket.StorageDealActive {
+		log.Infof("pending deal for proposal cid %s isn't active yet, deleting pending deal", dr.DealInfo.ProposalCid.String())
+		deletePending()
+	} else {
+		di, err := fromLotusDealInfo(ctx, m.api, info)
+		if err != nil {
+			log.Errorf("converting proposal cid %s from lotus deal info: %v", dr.DealInfo.ProposalCid.String(), err)
+			deletePending()
+			return
+		}
+		record := DealRecord{
+			Addr:     dr.Addr,
+			Time:     time.Now().Unix(), // Note: This can be much later in time than the deal actually became active on chain
+			DealInfo: di,
+			Pending:  false,
+		}
+		if err := m.store.putFinalDeal(record); err != nil {
+			log.Errorf("storing propoosal cid %s deal record: %v", dr.DealInfo.ProposalCid.String(), err)
+		}
+	}
+}
+
+func (m *Module) eventuallyFinalizeDeal(dr DealRecord, timeout time.Duration) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	updates, err := m.Watch(ctx, []cid.Cid{dr.DealInfo.ProposalCid})
+	if err != nil {
+		log.Errorf("watching proposal cid %s: %v", dr.DealInfo.ProposalCid.String(), err)
+		return
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			log.Infof("watching proposal cid %s timed out, deleting pending deal", dr.DealInfo.ProposalCid.String())
+			if err := m.store.deletePendingDeal(dr.DealInfo.ProposalCid); err != nil {
+				log.Errorf("deleting pending deal: %v", err)
+			}
+			return
+		case info, ok := <-updates:
+			if !ok {
+				log.Errorf("updates channel unexpectedly closed for proposal cid: %s", dr.DealInfo.ProposalCid.String(), err)
+				if err := m.store.deletePendingDeal(dr.DealInfo.ProposalCid); err != nil {
+					log.Errorf("deleting pending deal: %v", err)
+				}
+				return
+			}
+			if info.StateID == storagemarket.StorageDealActive {
+				record := DealRecord{
+					Addr:     dr.Addr,
+					Time:     time.Now().Unix(),
+					DealInfo: info,
+					Pending:  false,
+				}
+				log.Infof("proposal cid %s is active, storing deal record", info.ProposalCid.String())
+				if err := m.store.putFinalDeal(record); err != nil {
+					log.Errorf("storing propoosal cid %s deal record: %v", info.ProposalCid.String(), err)
+				}
+				return
+			} else if info.StateID == storagemarket.StorageDealProposalNotFound ||
+				info.StateID == storagemarket.StorageDealProposalRejected ||
+				info.StateID == storagemarket.StorageDealFailing ||
+				info.StateID == storagemarket.StorageDealNotFound {
+				log.Infof("proposal cid %s failed with state %s, deleting pending deal", info.ProposalCid.String(), storagemarket.DealStates[info.StateID])
+				if err := m.store.deletePendingDeal(info.ProposalCid); err != nil {
+					log.Errorf("deleting pending deal: %v", err)
+				}
+				return
+			}
+		}
+	}
+}
+
+func (m *Module) recordRetrieval(addr string, offer api.QueryOffer) {
```

Comment:
Recording retrieval deals is simple for now because we don't have and pending or final state. They are just logged.

Reply:
I could imagine some future version where there is a pending or intermediate state, but that would probably rely on underlying payment channels, and all of that isn't clear to me right now. So keeping it simple for now.